litestar/di.py

```diff
@@ -50,11 +55,14 @@ def __init__(
 
         self.dependency = Ref["AnyCallable"](dependency)
         self.has_sync_callable = isclass(dependency) or not is_async_callable(dependency)
-        if self.has_sync_callable and sync_to_thread is None:
+        if self.has_sync_callable and sync_to_thread is None and not is_sync_or_async_generator(dependency):
             warn_implicit_sync_to_thread(dependency, stacklevel=3)
 
-        if not self.has_sync_callable and sync_to_thread is not None:
-            warn_sync_to_thread_with_async_callable(dependency, stacklevel=3)
+        if sync_to_thread is not None:
+            if is_sync_or_async_generator(dependency):
+                warn_sync_to_thread_with_generator(dependency, stacklevel=3)
+            elif not self.has_sync_callable:
+                warn_sync_to_thread_with_async_callable(dependency, stacklevel=3)  # pyright: ignore
```

Comment:
I couldn't work out how to make pyright realize that `dependency` could not be `AnyGenerator` here.. :cry:

Reply:
Isn't it right in that it can be? It's only narrowed to be a non-generator here:

https://github.com/litestar-org/litestar/blob/7a00aa71e08df7e5728e71aeff976a1ec1055714/litestar/di.py#L58

Reply:
yeah - but shouldn't this narrow the generators away for the `else` branch:
https://github.com/litestar-org/litestar/blob/7a00aa71e08df7e5728e71aeff976a1ec1055714/litestar/di.py#L62

Reply:
Uhm, yes. It should. I just missed that :grimacing: